### ISSUE TITLE
wish: -str option for ocaml

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#5232]
**Original bug ID:** 5232
**Reporter:** @ygrek
**Status:** confirmed (set by @gasche on 2013-07-29T05:15:41Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Version:** 3.12.0
**Category:** toplevel
**Tags:** junior_job
**Has duplicate:** #6631
**Related to:** #2757 #6071
**Monitored by:** @gasche @ygrek

## Bug description

Sometimes it is useful to be able to run simple oneliner ocaml code without writing the code previously to file. Option -str analoguos to one of camlp4 is wanted.

**Comment by vicuna** [comment]
**Comment author:** @gasche

I think that this feature doesn't really work:
- the compilers (ocamlc and ocamlopt) need the filename to be used as a compilation unit name, so simply passing code in -str wouldn't work (except if this is only intended for parsing/typing with option -i?)
- the toplevel already makes it easy to work with short pieces of code (in particular `cat "foo;;" | ocaml`), so I'm not sure an -str option to ocaml would make much sense.

ygrek, could you be more specific on the use-cases and semantics you had in mind?

**Comment by vicuna** [comment]
**Comment author:** @ygrek

yes, this was intended exactly as a shorthand for toplevel, but with the following implied features : be usable as script runner, i.e. no extra toplevel output (prompts and types), stdin available for input (with toplevel it results in End_of_file), no reading of ocamlinit (not sure about this one). The intended use-case - running some automated utility tasks (inside project directory) where the actual code is available in a cma or ml file and one-liner is needed to actually call it or run with specific parameters.

**Comment by vicuna** [comment]
**Comment author:** @gasche

Thanks for the additional information. That seems reasonable. I think it's also reasonable to not run the .ocamlinit, because the "-stdin" option already doesn't do that. In fact I suspect they should both read .ocamlinit, and have a -noinit option (can be obtained as -init /dev/null right now), but that's question that can be handled independently (in fact it is an existing PR: #6071).

**Comment by vicuna** [comment]
**Comment author:** kyoDralliam

Seems related to #0006631

**Comment by vicuna** [comment]
**Comment author:** @gasche

Indeed; kyoDralliam, could you repost your patch here, using `-str` rather than `-e`?

PS: in case the patch is merged, the name you use in your .patch would be used as attribution information.

**Comment by vicuna** [comment]
**Comment author:** @mshinwell

@gasche Could you take care of these two issues?

**Comment by vicuna** [comment]
**Comment author:** @gasche

The patch is not good enough (for example it creates a temporary file but never deletes it, littering the user's filesystem with useless source files) and I had no feedback/return from the contributor. I can't spend time making a better patch right now, so as far as I am concerned the feature request is still standing, and any user's contribution is welcome.

**Comment by isaacv-pl** [comment]
Hi, I think I'm gonna work on this for my first contribution. Where in the toplevel folder would I find what I need to edit to support this option? Also an example of some usage behavior would be nice to clarify what I would need to do. Thanks! Sorry for all the questions.

**Comment by gasche** [comment]
@notemaster sorry for not answering this. Here would be an usage example:

```shell
# the following currently works
$ echo 'print_endline "Hello World";;' | ocaml -stdin
Hello World

# the following would requires this feature
$ ocaml -str 'print_endline "Hello World";;'
Hello World
```

**Comment by pkhry** [comment]
can i pick up this issue?

**Comment by dbuenzli** [comment]
Language interpreters usually use `-c` (e.g. sh, python) or `-e` (e.g. perl) for that option. Could we maybe use that aswell ? 

I can see `ocaml -c` as being confusing for `ocamlc` users, but `ocaml -e` looks fine no ?

**Comment by pkhry** [comment]
maybe -e, --eval ?

**Comment by gasche** [comment]
I agree that `-e` is fine.

**Comment by Octachron** [comment]
Fixed by #10438 .
